state-sync/storage-service/server/src/handler.rs

```diff
@@ -400,10 +406,8 @@ impl<T: StorageReaderInterface> Handler<T> {
         let storage_response = StorageServiceResponse::new(data_response, request.use_compression)?;
 
         // Cache the response before returning
-        let _ = self
-            .lru_response_cache
-            .lock()
-            .put(request.clone(), storage_response.clone());
+        self.lru_response_cache
```

Comment:
This is interesting. I was expecting some kind of [get_with](https://docs.rs/moka/latest/moka/sync/struct.Cache.html#method.get_with) but looks like moka-mini doesn't have that.

I'm wondering if this matters though. For any request, would the response possibly change with additional data? For example-

So a question about the original design here. Could this happen, and is it ok?:
A request comes in with a large end_index, and but there is only data up to end_index - 10 at this point. Then any subsequent requests with the same start_index, end_index will always return only up to end_index - 10.

Then additionally for this change, is it ok if the same start_index, end_index can return different answers, and the first returns end_index - 10 but the second only returns end_index - 20, and all subsequent calls of the same request also return end_index - 20?

Reply:
> I was expecting some kind of [get_with] but looks like moka-mini doesn't have that.

Yeah, it's kinda annoying, but not a deal breaker (see below). 😄 

The high-level design is essentially that it's okay for two identical requests to race, miss the cache, fetch from storage and insert them into the cache. The work is duplicated, but this is okay because: (i) any requests that are identical will always have identical responses (see below); (ii) in the current `LRUCache` implementation, there's a single `Mutex`, so we don't want to block the entire cache (and all other requests) while waiting to fetch the data for one request; and (iii) even in the new design, it appears we don't have the necessary interfaces to block only the identical requests -- but this is ultimately okay.

> A request comes in with a large end_index, and but there is only data up to end_index - 10 at this point.

This isn't possible -- any request that explicitly requests more data than the node has is invalid and will be dropped by the request moderator.

> is it ok if the same start_index, end_index can return different answers, and the first returns end_index - 10 but the second only returns end_index - 20, and all subsequent calls of the same request also return end_index - 20?

I don't think this is possible. Any identical requests will get the same data. The only edge case is that they get less data than they requested (i.e., they requested 10k items, but the max is 1k, so we truncate). But, even in this case, the truncation rule is static and deterministic (i.e., defined by the configs), so both requests will be truncated identically.

Taking a step back though, even if what you were saying is possible (which I'm confident it's not), this still wouldn't be a big issue because the client-side already has to deal with truncated data, so the clients would identify the missing data pieces and re-request them 😄 

Does this make sense? Have I understood the question?

Reply:
Thanks @JoshLind your explanation makes sense.

Just FYI, in my mempool dissemination branch, I use `moka` (not `mini-moka`) and it provides `get_with*`. But it's also a larger dependency.